### ISSUE TITLE
4.15.9 Release checklist

### ISSUE THREAD

**Issue by compulim** [microsoft/BotFramework-WebChat#4832]
# Checklist

## Build

1. [x] ~Bump MockBot to latest Bot Framework SDK release~ (not needed for patch release)
   - https://www.npmjs.com/package/botbuilder/v/latest
1. [x] Bump `botframework-directlinejs` to `0.15.4`
1. [x] Bump to `4.15.9`
   - [x] Update `CHANGELOG.md` to mark specific changes in `4.15.9`
   - [x] Run `npm version --no-git-tag-version 4.15.9`
   - [x] Merged into `main`, the PR number is #4834
   - Commit is `3eeebe4`
   - Do not merge any other unrelated changes after this PR. Any other PR merged, will need to be re-tested
1. [x] Run [daily pipeline](https://fuselabs.visualstudio.com/BotFramework-WebChat/_build?definitionId=498) manually, set "generate release version number" to `true`
   - (This will not push to NPM or CDN)
   - Pipeline name is `BotFramework-WebChat-daily`
   - The build number is [`365887`](https://fuselabs.visualstudio.com/BotFramework-WebChat/_build/results?buildId=365887&view=results) and commit is `3eeebe49`
1. [x] Wait for [`WebChat-release-testing` pipeline](https://fuselabs.visualstudio.com/BotFramework-WebChat/_release?_a=releases&view=mine&definitionId=17) to complete
   - Pipeline name is `Push-Release-Testing-to-GitHub-Pages`
   - The release ID is [`490`](https://fuselabs.visualstudio.com/BotFramework-WebChat/_releaseProgress?_a=release-pipeline-progress&releaseId=3172)
1. [x] Check component governance and make sure there are no high/critical related to code under `/packages/` folder
   - There could be some for projects under `/samples/` folder, as they are pointing to previous version of Web Chat
1. [x] Add manual tests to `WebChat-release-testing` as needed

## Test

> The test should run against the build artifacts from Azure Pipelines.

1. [x] Manual testing on major browsers using `webchat-release-testing`
   - [x] Before starting testing, update all the browser version to latest
   - [x] Chrome 116.0.5845.111
   - [x] Edge 118.0.2060.0
   - [x] Firefox 116.0.3
   - [x] IE11 (Windows 11 22H2 23531.1001)
   - [x] macOS Safari 16.5 (18615.2.9.11.4)
   - [x] iOS Safari 16.6 (20G75)
   - [x] iPadOS Safari 16.6 (20G75)
   - [x] Android Chrome 116.0.5845.92
1. [x] Test specific fixes related to `4.15.9` and previous releases

## Release

1. [x] Make sure you are on `main` ~or `qfe`~ branch, run `git status` to check
1. [x] `git pull`
1. [x] Verify `/package.json`, `/package-lock.json`, and `CHANGELOG.md` has a version of `4.15.9`
1. [x] `git log`
   - Verify the latest commit is `3eeebe4`
1. [x] `git tag v4.15.9`
1. [x] `git push -u upstream v4.15.9`
   - You do not need to kick off a build again, use the previous build
1. [x] Create a new GitHub release
   - [x] Copy entries from `CHANGELOG.md`
   - [x] Subresource Integrity can be generated by
      - From local: `for file in $(ls *.js); do echo $file $(cat $file | openssl dgst -sha384 -binary | openssl base64 -A); done`
      - From CDN: `curl -H 'Accept-Encoding: gzip' https://cdn.botframework.com/botframework-webchat/4.15.9/webchat.js | gunzip - | openssl dgst -sha384 -binary | openssl base64 -A`
   - [x] Attach assets including 3 JS files, `stats.json` and 5 tarballs
      - You can copy the artifacts from `webchat-release-testing/drops`
      - Tarballs download from npmjs
         ```
         curl -LO https://registry.npmjs.org/botframework-directlinespeech-sdk/-/botframework-directlinespeech-sdk-4.15.9.tgz
         curl -LO https://registry.npmjs.org/botframework-webchat/-/botframework-webchat-4.15.9.tgz
         curl -LO https://registry.npmjs.org/botframework-webchat-core/-/botframework-webchat-core-4.15.9.tgz
         curl -LO https://registry.npmjs.org/botframework-webchat-api/-/botframework-webchat-api-4.15.9.tgz
         curl -LO https://registry.npmjs.org/botframework-webchat-component/-/botframework-webchat-component-4.15.9.tgz
         ```
1. [x] Kick off release to NPM
   - Release name is `[[PROD]]Push-WebChat-to-npmjs`
   - The build number is [365887](https://fuselabs.visualstudio.com/BotFramework-WebChat/_build/results?buildId=xxxxxx&view=results) release number is [`47`](https://fuselabs.visualstudio.com/BotFramework-WebChat/_releaseProgress?_a=release-pipeline-progress&releaseId=3173) and commit is `3eeebe49`
   - [x] Verify package content then click Resume
   - [x] Retain the release indefinitely
1. [ ] Kick off release to CDN (cutoff at 10 PM PST, Sun-Wed only)
   1. [x] Prepare the message for approval
      - [x] ~If there are any breaking changes, explain in the email if it will affect any customers~
      - Release name is `[[PROD]]Push-WebChat-to-Prod-CDN-with-approval`
      - The build number is [`365887`](https://fuselabs.visualstudio.com/BotFramework-WebChat/_build/results?buildId=365887&view=results), release number is [`51`](https://fuselabs.visualstudio.com/BotFramework-WebChat/_releaseProgress?_a=release-pipeline-progress&releaseId=3210) and commit is `3eeebe49`
      - Script build number is [`320590`](https://fuselabs.visualstudio.com/BotFramework-WebChat/_build/results?buildId=320590&view=results) (this is fixed)
   1. [x] Send message to approvers
   1. [ ] Retain the build indefinitely

## Post-release verification - complete within 30 minutes after release to NPM

- [x] Test using `webchat-release-testing`
   1. [x] Clone https://github.com/corinagum/WebChat-release-testing/
   1. [x] `01.create-react-app`
      1. [x] Nuke `01.create-react-app/node_modules`
      1. [x] `npm install`
      1. [x] `npm install botframework-webchat@4.15.9` (just install the bundle package)
      1. [x] `npm run build`
   1. [x] Others
      -  Using script tags from https://github.com/microsoft/BotFramework-WebChat/releases/tag/v4.15.9, with subresource integrity
         ```html
         <script
            crossorigin="anonymous"
            integrity="sha384-kr+6QqBRD45fWiraMYzQdm0CpHzmKMtNk0gRLBVqa379eQdjCHNnm8jfmAhiOQnI"
            src="https://cdn.botframework.com/botframework-webchat/4.15.9/webchat.js"
         ></script>

         <script
            crossorigin="anonymous"
            integrity="sha384-04jpTOxuj3fbYmyDC2/jU3rdwhoC4OWnNbo6zh6RNEq+qvzNggRWDzdN788GnXs6"
            src="https://cdn.botframework.com/botframework-webchat/4.15.9/webchat-es5.js"
         ></script>

         <script
            crossorigin="anonymous"
            integrity="sha384-nCB7vnhkKK/wLHWq8fqOaZM5YlsR0X/6FcS31MOnl7aYmMB+qo3qPp0hacahdAbN"
            src="https://cdn.botframework.com/botframework-webchat/4.15.9/webchat-minimal.js"
         ></script>
         ```
   1. [x] `npx serve` (at repo root)
   1. [x] Go to http://localhost:5000/ to test

## Notification to interested parties

- [x] Update partner page on Adaptive Cards doc
   - https://docs.microsoft.com/en-us/adaptive-cards/resources/partners
   - PR at https://github.com/MicrosoftDocs/AdaptiveCards/pull/524
- [x] Notify related parties for the following fixes
   - [x] Azure Bot Services
   - [x] Omnichannel
   - [x] Power Virtual Agents
- [x] ~Update root README.md with feature notes -- Note: PR will be combined with post-release checklist PR~
   
---

# Post-release checklist

These are chores that we should do before starting the cycle to reduce ripple effects if we do it in mid-cycle.

Tips: 

- Clean your repo before start
- Remove `node_modules` from all folder
   - `git clean -fdx`
- Never delete `package-lock.json`
- If you mess it up, tableflip and redo
- In `component/package.json`
   - Remove reference to `botframework-webchat-core` by hand-modifying `package.json`
   - Then, `npm install` (symlinks will be broken afterward)
   - Then, add those references back by hand-modifying `package.json`
   - This also applies for other packages with similar dependencies/symlinks
   - To build afterward, do tableflip to rebuild those symlinks

## Applies to all releases

> This list should be copied to versions in the future.

- [x] ~If on QFE branch, make sure `CHANGELOG.md` and version number bump is cherry-picked to `main`~
   - [x] ~`git checkout main`~
   - [x] ~`git cherry-pick XXX` (the commitish for bumping version number and `CHANGELOG.md`)~
- [x] ~If needed, correct the date for 4.15.9 in `CHANGELOG.md` in PR #XXX~
   - ~There could be last minute fixes that could push the planned date later than the one in `CHANGELOG.md`~
- [x] Bump `package.json` to `4.15.10-0` in PR #4838
   - Run `npm version prepatch --no-git-tag-version`
- [ ] Update [`servicingPlan.json`](https://github.com/microsoft/BotFramework-WebChat/blob/main/packages/embed/servicingPlan.json) in PR #XXX
   - Add deprecation notes for previous versions
   - Subresource integrity hash from https://github.com/microsoft/BotFramework-WebChat/releases/tag/v4.15.9
- [x] ~Update all samples to use `4.15.9` in PR #XXXX~
   - Some samples are pointing to GitHub Releases because the sample need new features from daily build
   - Search "https://github.com/microsoft/BotFramework-WebChat/releases/download/"
      - And replace with "https://cdn.botframework.com/botframework-webchat/latest/"
- [x] ~Clean up unnecessary branch on official repo~
- [x] Understand production-hitting vulnerabilities
   - [x] Create a new folder
   - [x] Run `npm init` with default values
   - [x] Run `npm install botframework-webchat@4.15.9`
   - [x] Look at the result and see if there are any production-hitting vulnerabilities, investigate if needed
      - No vulnerabilities found
      - ![image](https://github.com/microsoft/BotFramework-WebChat/assets/1622400/e2146e65-bcbb-490e-aa0e-a00575ed9cfd)
- [x] Bump in Power Virtual Agents
- [x] Update [accessibility conformance report](https://cela-ra-aat-prod.azurewebsites.net/#/workflow/e4fd918a-9ef5-4c0c-9d93-98e2ae13396a?tab=default)
   - For any accessibility bugs resolved, remove them from external notes
   - Update version number and publish date

### Bump all dependencies to latest version

In PR #4647, we are bumping most dependencies to latest version.

> After bumping, if a package broke compatibility, we should investigate:
>
> - Upgrade our code to use the latest package if possible, otherwise;
> - Add it to `package.json/skipBump` to prevent bumping deliberately:
>    - Skipping bump incur unpredictable technical debts, say, security issue found in the unsupported version, causing us slow to react
>    - Plausible reasons (non-exhaustive):
>       - Package is not ES5;
>       - Package is ESM and requires the whole dependency chain to be upgraded, however, it is technically impossible (unrelated to cost).

- [ ] Run `npm run bump`
- [ ] Run `npm audit fix` to make sure everything is fixed
- [ ] Test under IE11 to make sure all dependencies are working
- [ ] List steps to verify bumping `microsoft-cognitiveservices-speech-sdk`

### Bump Docker image

> The Docker image can be found at root `docker-compose.yml` and `Dockerfile*`.

- [ ] Docker container for headless Chrome in [PR #XXXX](https://github.com/microsoft/BotFramework-WebChat/pull/XXXX)
   - Bumped to Chrome 110 and Selenium Hub 4.8.1

**Comment by compulim** [comment]
Obsoleted by #4954.
